document and clean up MaybeUpdateMempoolForReorg

glozow opened the pull request:
followup to #23683, addressing https://github.com/bitcoin/bitcoin/pull/23683#issuecomment-989741186

glozow committed:
document and clean up MaybeUpdateMempoolForReorg

Co-authored-by: John Newbery <john@johnnewbery.com>

glozow commented:
cc @jnewbery

MarcoFalke commented on src/validation.cpp:
```diff
@@ -340,21 +340,29 @@ void CChainState::MaybeUpdateMempoolForReorg(
+        // So it's critical that we remove the tx and not depend on the LockPoints.
+        if (!CheckSequenceLocks(m_chain.Tip(), view_mempool, tx, flags, &lp, validLP)) {
+            return true;
+        } else if (!validLP) {
```
```suggestion
        }  if (!validLP) {
```

nit: no need for else. I find it confusing in early return statements.

jnewbery replied:
See rationale here: https://github.com/bitcoin/bitcoin/pull/23683#discussion_r772879178 , but you're right that they've equivalent, so both are fine.

hebasto replied:
Why this branch--`if (!validLP)`--has been moved before `if (it->GetSpendsCoinbase())` ?

jnewbery replied:
Because logically it's connected to the previous if statement - if `CheckSequenceLocks()` returns false then this filter function fails immediately. If it returns true, then the lockpoints may be invalid so they should be updated.

Functionally this is equivalent, but it groups the related functionality together.

glozow replied:
Clarified the comments so it should be more clear to a first-time reader why the branches are linked

hebasto commented:
Related #23897.

hebasto reviewed (commented):
Concept ACK.

What about
```diff
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -348,11 +348,11 @@ void CChainState::MaybeUpdateMempoolForReorg(
         AssertLockHeld(m_mempool->cs);
         AssertLockHeld(::cs_main);
         const CTransaction& tx = it->GetTx();
+        if (!CheckFinalTx(m_chain.Tip(), tx, flags)) return true;
         LockPoints lp = it->GetLockPoints();
         const bool validLP{TestLockPointValidity(m_chain, lp)};
         CCoinsViewMemPool view_mempool(&CoinsTip(), *m_mempool);
         // The transaction must be final.
-        if (!CheckFinalTx(m_chain.Tip(), tx, flags)) return true;
         // Note if CheckSequenceLocks fails the LockPoints may still be invalid
         // So it's critical that we remove the tx and not depend on the LockPoints.
         if (!CheckSequenceLocks(m_chain.Tip(), view_mempool, tx, flags, &lp, validLP)) {
```
?

instagibbs commented on src/validation.cpp:
```diff
@@ -340,21 +340,29 @@ void CChainState::MaybeUpdateMempoolForReorg(
+    // Check whether the transaction is still final and, if it spends a coinbase output, mature.
+    // Update the entry's cached LockPoints if needed. If this returns true, the tx would be invalid
+    // in the next block, so the mempool must remove this entry and all of its descendants.
+    const auto filter_final_and_mature = [this, flags=STANDARD_LOCKTIME_VERIFY_FLAGS](CTxMemPool::txiter it)
```
nit: this sounds like we're getting rid of final/mature txs.

a little longer but `filter_for_final_and_mature` makes it less ambiguous to me

glozow replied:
I've updated the comments to make it more clear that this is meant to be a predicate for filtering, so returning `true` for something we want to filter out is natural.

instagibbs commented on src/validation.cpp:
```diff
@@ -340,21 +340,29 @@ void CChainState::MaybeUpdateMempoolForReorg(
-        } else if (it->GetSpendsCoinbase()) {
+        // The transaction must be final.
+        if (!CheckFinalTx(m_chain.Tip(), tx, flags)) return true;
+        // Note if CheckSequenceLocks fails the LockPoints may still be invalid
```
while we're here: I'm not sure what this comment block is saying.

glozow replied:
Better now?

instagibbs replied:
yes the comment you left actually matches my understanding of the code

glozow committed:
Replace `struct update_lock_points` with lambda

No behavior change.
This code was introduced in 5add7a7 before we required C++11, which is
why the struct was needed. As we are now using more modern C++ and this
is the only place where lockpoints are updated for mempool entries, it
is more idiomatic to call `modify` with a lambda.

Co-authored-by: Hennadii Stepanov <32963518+hebasto@users.noreply.github.com>

glozow commented:
Sorry for the delay! I've addressed @hebasto and @instagibbs comments and grouped in @hebasto's commit from #23958.

glozow commented on src/validation.cpp:
```diff
@@ -363,18 +371,16 @@ void CChainState::MaybeUpdateMempoolForReorg(
 
     // We also need to remove any now-immature transactions
-    m_mempool->removeForReorg(m_chain, check_final_and_mature);
+    m_mempool->removeForReorg(m_chain, filter_final_and_mature);
```
Fixed

instagibbs reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/23976/commits/283936b10d2e218726648bc9cd640d56d6922bb2

hebasto reviewed (commented):
It looks like some changes that belong to the second commit--a replacing `update_lock_points` with a lambda--go to the first commit accidentally, no?

glozow commented:
> It looks like some changes that belong to the second commit--a replacing `update_lock_points` with a lambda--go to the first commit accidentally, no?

Oops! Good catch, fixed.

hebasto reviewed (approved):
ACK e177fcab3831b6d259da5164cabedcc9e78f6957, I have reviewed the code and it looks OK, I agree it can be merged.

instagibbs reviewed (approved):
ACK e177fcab3831b6d259da5164cabedcc9e78f6957

MarcoFalke commented on src/validation.cpp:
```diff
@@ -372,18 +388,16 @@ void CChainState::MaybeUpdateMempoolForReorg(
                 assert(!coin.IsSpent());
                 const auto mempool_spend_height{m_chain.Tip()->nHeight + 1};
                 if (coin.IsSpent() || (coin.IsCoinBase() && mempool_spend_height - coin.nHeight < COINBASE_MATURITY)) {
```
nit: any reason to check for `coin.IsSpent()`, given the assertion?

maflcko reviewed (approved):
Approach ACK e177fcab3831b6d259da5164cabedcc9e78f6957 😶

Did not review the last commit, but given the assert added in b4adc5ad6769e4a5a6179dfff271cd4c9dc47a5b it should be correct.

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Approach ACK e177fcab3831b6d259da5164cabedcc9e78f6957 😶

Did not review the last commit, but given the assert added in b4adc5ad6769e4a5a6179dfff271cd4c9dc47a5b it should be correct.
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUj/Ugv+N4lEh7A93jSO5imukN2scMEs6r7j7KNiTudrany2XPrSbTvJQgl+q46K
j5UPWvQCzJ0M+NpenYE0Eiefl51eZGRo+4rT06GlQJ/jNyZkutbVmbbNvspM19gg
lsnpx7pfaLR9MFAmIZQ3xz2y5zu2adewPtFrTAfKbxfkZbkR17+EayICGvXRked9
u/UGf/F/XmA/Yopnfn4jAHpPUlvvgKFxOaRFI76302IDY6D9XgzTAg2cjLBlUuy9
3KCf5UGPJ2fVYFcpf4OZdsGi+vzo17+UGOK0dcYb/wLqfr2VH/mZPL5o/jo2l3d4
EcGvDAzYRrM3tvRp9pFDO0O27avyvMw106RI/u1/TyqKX4apcblX954bAugIBXuN
ZgAed34pqaHeRqm9R1WdY1O270pehbUtv+NKWu6awEZ+j79vwF8UVwOModkepwbd
Fp0c+f4aEAzHLEEdpMR2l+XUrF97ZviaVNGXX06bfuIpSsw7MpR3K98SMmLxVyn9
CVsQAgjV
=YL6U
-----END PGP SIGNATURE-----
```


</details>
